In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
from sklearn.preprocessing import MinMaxScaler

In [3]:
data=pd.read_csv('/home/iswar/Downloads/titanic_dataset (2).csv')
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [8]:
data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [9]:
data.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

# Data Preprocessing

In [10]:
data1=data
data1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

The data from column 'cabin' is mostly missing. So this column is dropped.

In [11]:
data1.drop('Cabin',axis=1,inplace=True)

In [15]:
data1["Age"] = data1["Age"].fillna(data1["Age"].mean())
data1['Embarked']= data1['Embarked'].fillna(data1['Embarked'].mode()[0])

In [16]:
data1.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

In [17]:
data.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

Handling outliers

In [20]:
data1.drop(['PassengerId','Name','Ticket'],axis=1,inplace=True)

In [21]:
data1.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0         0       3    male  22.0      1      0   7.2500        S
1         1       1  female  38.0      1      0  71.2833        C
2         1       3  female  26.0      0      0   7.9250        S
3         1       1  female  35.0      1      0  53.1000        S
4         0       3    male  35.0      0      0   8.0500        S

In [23]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
data1['Sex']= le.fit_transform(data1['Sex'])
data1['Embarked']= le.fit_transform(data1['Embarked'])

In [24]:
data1.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    1  22.0      1      0   7.2500         2
1         1       1    0  38.0      1      0  71.2833         0
2         1       3    0  26.0      0      0   7.9250         2
3         1       1    0  35.0      1      0  53.1000         2
4         0       3    1  35.0      0      0   8.0500         2

In [25]:
x=data1.drop('Survived',axis=1)
y=data1['Survived']

In [26]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=10)

In [27]:
from sklearn.preprocessing import StandardScaler
ss=StandardScaler()
x_train=ss.fit_transform(x_train)
x_test=ss.transform(x_test)

In [28]:
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score, precision_score,recall_score, f1_score

Naive Bayes Algorithm

In [32]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(x_train,y_train)
y_predgnb = gnb.predict(x_test)
y_train_predgnb= gnb.predict(x_train)
print('Accuracy Score is:', accuracy_score(y_test, y_predgnb))
print('Recall Score is:', recall_score(y_test, y_predgnb))
print('Precision Score:', precision_score(y_test, y_predgnb))
print('F1 score is:', f1_score(y_test, y_predgnb))
print("\n The confusion matrix is : \n", confusion_matrix(y_test, y_predgnb))

Accuracy Score is: 0.8022388059701493
Recall Score is: 0.7340425531914894
Precision Score: 0.711340206185567
F1 score is: 0.7225130890052357

 The confusion matrix is : 
 [[146  28]
 [ 25  69]]


Decision Tree Classifier

In [33]:
from sklearn.tree import DecisionTreeClassifier
dtc= DecisionTreeClassifier()
dtc.fit(x_train,y_train)
y_preddtc= dtc.predict(x_test)
y_train_preddtc = dtc.predict(x_train)
print("Training Accuracy is : ", accuracy_score(y_train, y_train_preddtc))
print('Accuracy Score is:', accuracy_score(y_test, y_preddtc))
print('Recall Score is:', recall_score(y_test, y_preddtc))
print('Precision Score:', precision_score(y_test, y_preddtc))
print('F1 score is:', f1_score(y_test, y_preddtc))
print("\n The confusion matrix is : \n", confusion_matrix(y_test, y_preddtc))

Training Accuracy is :  0.9871589085072231
Accuracy Score is: 0.7723880597014925
Recall Score is: 0.6702127659574468
Precision Score: 0.6774193548387096
F1 score is: 0.6737967914438502

 The confusion matrix is : 
 [[144  30]
 [ 31  63]]


Random Forest Classifier

In [35]:
from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(x_train,y_train)
y_predrfc= rfc.predict(x_test)
y_train_predrfc= rfc.predict(x_train)

print("Training Accuracy is : ", accuracy_score(y_train, y_train_predrfc))
print('Accuracy Score is:', accuracy_score(y_test, y_predrfc))
print('Recall Score is:', recall_score(y_test, y_predrfc))
print('Precision Score:', precision_score(y_test, y_predrfc))
print('F1 score is:', f1_score(y_test, y_predrfc))
print("\n The confusion matrix is : \n", confusion_matrix(y_test, y_predrfc))

Training Accuracy is :  0.9871589085072231
Accuracy Score is: 0.8208955223880597
Recall Score is: 0.7553191489361702
Precision Score: 0.7395833333333334
F1 score is: 0.7473684210526316

 The confusion matrix is : 
 [[149  25]
 [ 23  71]]


Logistic Regression Model

In [36]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(x_train,y_train)
y_predlr = lr.predict(x_test)
y_train_predlr = lr.predict(x_train)
print("Training Accuracy is : ", accuracy_score(y_train, y_train_predlr))
print('Accuracy Score is:', accuracy_score(y_test, y_predlr))
print('Recall Score is:', recall_score(y_test, y_predlr))
print('Precision Score:', precision_score(y_test, y_predlr))
print('F1 score is:', f1_score(y_test, y_predlr))
print("\n The confusion matrix is : \n", confusion_matrix(y_test, y_predlr))

Training Accuracy is :  0.8025682182985554
Accuracy Score is: 0.8097014925373134
Recall Score is: 0.7021276595744681
Precision Score: 0.7415730337078652
F1 score is: 0.7213114754098361

 The confusion matrix is : 
 [[151  23]
 [ 28  66]]


KNN Model

In [37]:
from sklearn.neighbors import KNeighborsClassifier
knnc = KNeighborsClassifier(n_neighbors=4,metric = "minkowski") 
knnc.fit(x_train,y_train)
y_predknnc = knnc.predict(x_test)
y_train_predknnc = knnc.predict(x_train)
print("Training Accuracy is : ", accuracy_score(y_train, y_train_predknnc))
print('Accuracy Score is:', accuracy_score(y_test, y_predknnc))
print('Recall Score is:', recall_score(y_test, y_predknnc))
print('Precision Score:', precision_score(y_test, y_predknnc))
print('F1 score is:', f1_score(y_test, y_predknnc))
print("\n The confusion matrix is : \n", confusion_matrix(y_test, y_predknnc))

Training Accuracy is :  0.8475120385232745
Accuracy Score is: 0.8059701492537313
Recall Score is: 0.648936170212766
Precision Score: 0.7625
F1 score is: 0.7011494252873564

 The confusion matrix is : 
 [[155  19]
 [ 33  61]]


Comparing the accuracy scores given above, Random forest classifier has the highest accuracy with a score of 0.8208955223880597